x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java

```diff
@@ -1891,6 +1891,17 @@ public void testLookup() {
         assertThat(e.getMessage(), containsString("LOOKUP not yet supported"));
     }
 
+    public void testImplicitCasting() {
+        var e = expectThrows(VerificationException.class, () -> analyze("""
+             from test | eval x = concat("2024", "-04", "-01") + 1 day
```

Comment:
Maybe we could add one _last_ test with:
```suggestion
| eval x = to_string(null) - 1 day
```

I realize I'm getting further into paranoid territory :D However, this would demonstrate that implicit casting happens really only when there's a literal string, rather than a string expression.

(I see that the analyzer correctly complains about `to_string(null)` being of type `keyword`, nice :) .)

Reply:
> Maybe we could add one _last_ test with:
> 
> ```
> | eval x = to_string(null) - 1 day
> ```
> 
> I realize I'm getting further into paranoid territory :D However, this would demonstrate that implicit casting happens really only when there's a literal string, rather than a string expression.
> 
> (I see that the analyzer correctly complains about `to_string(null)` being of type `keyword`, nice :) .)



> Maybe we could add one _last_ test with:
> 
> ```
> | eval x = to_string(null) - 1 day
> ```
> 
> I realize I'm getting further into paranoid territory :D However, this would demonstrate that implicit casting happens really only when there's a literal string, rather than a string expression.
> 
> (I see that the analyzer correctly complains about `to_string(null)` being of type `keyword`, nice :) .)

Yes it works for string literals only for now, we may open it up for expressions/functions that return a string in the future. I'll add the suggested test.